osu.Game/Screens/Select/BeatmapDetails.cs

```diff
@@ -0,0 +1,430 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using OpenTK;
+using OpenTK.Graphics;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Database;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using System.Globalization;
+using System.Linq;
+
+namespace osu.Game.Screens.Select
+{
+    public class BeatmapDetails : Container
+    {
+        private readonly MetadataSegment description;
+        private readonly MetadataSegment source;
+        private readonly MetadataSegment tags;
+
+        private readonly DifficultyRow circleSize;
+        private readonly DifficultyRow drainRate;
+        private readonly DifficultyRow overallDifficulty;
+        private readonly DifficultyRow approachRate;
+        private readonly DifficultyRow stars;
+
+        private readonly Container ratingsContainer;
+        private readonly Bar ratingsBar;
+        private readonly OsuSpriteText negativeRatings;
+        private readonly OsuSpriteText positiveRatings;
+        private readonly BarGraph ratingsGraph;
+
+        private readonly FillFlowContainer retryFailContainer;
+        private readonly BarGraph retryGraph;
+        private readonly BarGraph failGraph;
+
+        private BeatmapInfo beatmap;
+        public BeatmapInfo Beatmap
+        {
+            get
+            {
+                return beatmap;
+            }
+            set
+            {
+                beatmap = value;
+                if (beatmap == null) return;
+
+                description.Text = beatmap.Version;
+                source.Text = beatmap.Metadata.Source;
+                tags.Text = beatmap.Metadata.Tags;
+
+                circleSize.Value = beatmap.Difficulty.CircleSize;
+                drainRate.Value = beatmap.Difficulty.DrainRate;
+                overallDifficulty.Value = beatmap.Difficulty.OverallDifficulty;
+                approachRate.Value = beatmap.Difficulty.ApproachRate;
+                stars.Value = (float)beatmap.StarDifficulty;
+
+                if (beatmap.Metric?.Ratings.Count() == 10)

```

Comment:
you shouldn't be enforcing these here. such conditions should be enforced in the model if anywhere.

i would change the range calculations below (all the places you use `5`) to be `Count / 2` and change the if conditional to `Count > 0`.